### PR TITLE
Render button component using a home grown HTML DSL

### PR DESCRIPTION
**Pull request by TobiasBales**
We tried to use [phlex](https://github.com/joeldrapper/phlex) but there were roughly a million issues stopping us from adopting it (from it expecting to live in a rails like environment, to having non-skippable safety checks to add onclick handlers and a lot more).

So we decided to implement a minimal DSL that makes it a bit saner to build HTML

👍 or 👎?

paired with @CalvinWalzel
relates to #4

### CODE REVIEW

**Comment by Schwad** [comment]
This is really interesting. Thank you for sharing! Sorry that Phlex was giving you issues, maybe it's not the right tool for the job in our particular use case.

(CC'ing the creator @joeldrapper - We're trying to spit out raw HTML in one part of our app and something I'd asked the team was to consider Phlex)

I'm slightly nervous about merging our own DSL... but since we are doing a hackathon I'm open to new ideas generally!

I wonder if [markaby](https://github.com/markaby/markaby#using-markaby-as-a-ruby-class-) might be one option because of how it supports this syntax and a `to_s` as well.

What do you think about that? Also open to hear feedback and thoughts from other hackathoners!

**Comment by TobiasBales** [comment]
> 

we looked at libraries but did not see markaby - we'll see if it works as we'd like it to

**Comment by TobiasBales** [comment]
> > 
> 
> we looked at libraries but did not see markaby - we'll see if it works as we'd like it to

We looked at it and the issue is that it uses [instance eval](https://github.com/markaby/markaby#label-A+Note+About+instance_eval) which means we can't access attributes and methods from the calling class (we explicitly worked around that in our design) and we couldn't see a way to use markaby like that - if we missed it we'd love to know @Schwad

File: lib/scarpe/html.rb [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+      @buffer
+    end
+
+    private
+
+    def p(*args, &block)
+      method_missing(:p, *args, &block)
+    end
+
+    def method_missing(name, *args, &block)
```
[the remark this answers is not in the document]
Comment (TobiasBales) [reply lib/scarpe/html.rb:26]: Done

**Comment by Schwad** [comment]
Thanks for the discussion and deep diving into this!

I'm tempted to accept this approach because it's non-blocking (E.g. someone can start with raw html and work backwards if it's tricky)

I think the feel and overall look is improved with this too 👍

**Comment by Schwad** [comment]
> Hey, Phlex core shouldn't have any dependencies on Rails. Were you using `phlex` or `phlex-rails`?
> 
> I don't have any plans to allow the `onclick` attribute at the moment as there are better ways to listen for events and `onclick` can be exploited for cross-site scripting. If this is a hackathon, you could monkey-patch `Phlex::HTML::EVENT_ATTRIBUTES = []` to disable the checks.

@joeldrapper thank you so much for dropping in and for your context here! Appreciate it and your work 🙇

**Review by Schwad — APPROVED** [review]

File: lib/scarpe/html.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
+module Scarpe
```
Comment (Schwad) [inline lib/scarpe/html.rb:1]: If we find ourselves using it a lot, definitely a gem candidate at some point

**Comment by Schwad** [comment]
I'll update some of the tasklists in relation to this

**Pull request merged**